network/transport/v2/handlers.go

```diff
@@ -151,10 +158,114 @@ func (p protocol) handleGossip(peer transport.Peer, msg *Gossip) error {
+
+	// check if response matches earlier request
+	if err = p.cMan.check(envelope); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	for _, tx := range msg.Transactions {
+		transactionRef := hash.FromSlice(tx.Hash)
+		transaction, err := dag.ParseTransaction(tx.Data)
+		if err != nil {
+			return fmt.Errorf("received transaction is invalid (peer=%s, ref=%s): %w", peer, transactionRef, err)
+		}
+
+		present, err := p.state.IsPresent(ctx, transactionRef)
+		if !present {
+			// TODO does this always trigger fetching missing payloads? (through observer on DAG) Prolly not for v2
+			if err = p.state.Add(ctx, transaction, tx.Payload); err != nil {
+				return fmt.Errorf("unable to add received transaction to DAG (tx=%s): %w", transaction.Ref(), err)
+			}
+		}
+	}
+
+	// TODO done is not called on the cMan since we don't know if we received all chunks
+
+	return nil
+}
+
+func (p *protocol) handleTransactionListQuery(peer transport.Peer, msg *TransactionListQuery) error {
+	requestedRefs := make([]hash.SHA256Hash, len(msg.Refs))
+	transactions := make([]*Transaction, 0)
+	unsorted := make([]dag.Transaction, 0)
+
+	cid, err := parseConversationID(msg.ConversationID)
+	if err != nil {
+		return err
+	}
+	// TODO convert to trace logging
+	log.Logger().Infof("handling transactionListQuery from peer (peer=%s, conversationID=%s)", peer.ID.String(), cid.String())
+
+	for i, refBytes := range msg.Refs {
+		requestedRefs[i] = hash.FromSlice(refBytes)
+	}
+
+	if len(requestedRefs) == 0 {
+		log.Logger().Warnf("peer sent request for 0 transactions (peer=%s)", peer.ID)
+		return nil
+	}
+
+	ctx := context.Background()
+
+	// first retrieve all transactions, this is needed to sort them on LC value
+	for _, ref := range requestedRefs {
+		transaction, err := p.state.GetTransaction(ctx, ref)
+		if err != nil {
+			return err
+		}
+		// If a transaction is not present, we stop any further transaction gathering.
+		if transaction == nil {
+			log.Logger().Warnf("peer requested transaction we don't have (peer=%s, node=%s, ref=%s)", peer.ID, peer.NodeDID.String(), ref.String())
+			break
+		}
+		unsorted = append(unsorted, transaction)
+	}
+
+	// now we sort on LC value
+	sort.Slice(unsorted, func(i, j int) bool {
+		return unsorted[i].Clock() <= unsorted[j].Clock()
+	})
+
+	for i, transaction := range unsorted {
+		networkTX := Transaction{
+			Hash: msg.Refs[i],
+			Data: transaction.Data(),
+		}
+
+		// do not add private TX payloads
+		if len(transaction.PAL()) == 0 {
+			payload, err := p.state.ReadPayload(ctx, transaction.PayloadHash())
+			if err != nil {
+				return err
+			}
+			// TODO we abort here as well, since there's no mechanism for missing payloads on public transactions in v2 protocol
+			if payload == nil {
+				log.Logger().Warnf("peer requested transaction with missing payload (peer=%s, node=%s, ref=%s)", peer.ID, peer.NodeDID.String(), transaction.Ref().String())
```

Comment:
should we sent an error back to the peer then, instead of just responding with what we collected up until now?

If you return an error from the handler, it is returned to the peer.

Reply:
Another node might have the payload, is this way nodes can share the burden. The sync with the other node will continue where this one was left. In a pure v2 protocol implementation, we wouldn't accept transactions without payloads anymore.